### ISSUE TITLE
Missing several m17n engines in new Mint 20/Cinnamon

### ISSUE THREAD

**Issue by foberle** [ibus/ibus-m17n#25]
Several of the m17n engines that I've used with iBus in the past are no longer available in Linux Mint 20, among them m17n:ar:kbd, m17n:el:kbd, m17n:he:kbd, and m17n:ko:han2 (Arabic, Greek, Hebrew, and Korean); the details are available in a posting to the Linux Mint forum at [https://forums.linuxmint.com/viewtopic.php?f=47&t=334843], so I'll try to keep this short. 

Although I am able to use PinYin for Chinese in Mint 20, a posting I found (http://www.pinyinjoe.com/linux/ubuntu-18-gnome-chinese-setup.htm) suggested that in newer versions of Ubuntu and derivatives, it is necessary to generate locales in order for certain engines to become available. I seemed to already have locale definitions for some of the keyboards I was missing, and adding others made no difference. So that doesn't seem relevant.

Not unexpectedly, there have been no responses to my posting on the Mint forum. I was wondering if anyone here knows where I can obtain these and, if so, what I need to do to add them to my system. Thanks for any pointers anyone can offer.

**Comment by mike-fabian** [comment]
Maybe because of this commit which blacklisted them?:

https://github.com/ibus/ibus-m17n/commit/289c8b738db5d424b796f0a9f79688d73b22280a

**Comment by mike-fabian** [comment]
For m17n:ar:kbd, m17n:el:kbd, m17n:he:kbd, and m17n:ko:han2 you could  consider using ibus-typing-booster instead of ibus-m17n.

https://github.com/mike-fabian/ibus-typing-booster

Ibus typing booster can use all the m17n input methods installed and doesn't blacklist any of them.

For Korean you can also use ibus-hangul, it is probably better for Korean than ibus-m17n.

But you can still use the m17n input methods for Korean with ibus-typing-booster which may be useful if you want to type many different  languages and want to reduce the switching between input methods. ibus-typing-booster can use several input methods at the same time which often makes switching unnecessary. 

https://repology.org/project/ibus-typing-booster/versions

doesn't show ibus-typing-booster packages for Linux Mint yet, but as Linux Mint is based on Ubuntu, maybe you can use the available Ubuntu packages?

**Comment by foberle** [comment]
Mike: 


Thanks so much for this reply. 


I have been typing on PCs in multiple scripts since I got my first Hercules RamFont graphics card back in the 1980s and built my own primitive Thai font. I thought I had been following technical developments fairly closely since then, so I was surprised that I had never encountered (or even heard of) the ibus typing booster project. 


Given its size, however, it will take me some time to read through all the material I discovered to make sure I understand how to use it before experimenting with it. 


I don't actually use Fedora on any of my machines, but I do generally keep up with posts on their magazine and found the article at https://fedoramagazine.org/master-typing-ibus-booster/ which I assume you've seen. 


So I guess I've got some exploring to do - hopefully I can find enough play time in the upcoming weeks. 


Again, thanks much for contacting me - it is much appreciated. 


Frank 


P.S. You were correct about the reason for the blacklisting - it just annoyed me because a) it didn't seem to be documented anywhere in any distro release notes and b) some of the engines that were blacklisted were (in my view) better than the replacements for a variety of reasons. 

----- Original Message -----

**Comment by mike-fabian** [comment]
> I don't actually use Fedora on any of my machines, but I do generally keep up with posts on their magazine and found the article at https://fedoramagazine.org/master-typing-ibus-booster/ which I assume you've seen.

This is a bit newer and maybe more Ubuntu specific:

https://blog.simos.info/using-ibus-typing-booster-for-predictive-typing-in-ubuntu-20-04-lts/

My documentation is here:

http://mike-fabian.github.io/ibus-typing-booster/documentation.html

**Comment by mike-fabian** [comment]
> b) some of the engines that were blacklisted were (in my view) better than the replacements for a variety of reasons.

Which ones and why?

**Comment by foberle** [comment]
Ahhh ... very tough questions and certainly personal choices based on my own usage. I suspect that most folks switch between two scripts/languages at a time, but since my usage relates to tutorials for handling data in multi-national databases, I may have single sentences containing words from several different languages. Databases for some companies with international reach may also have multiple scripts/languages intermingled. 



So, without "picking on" individual engines (which is probably unfair), the chief annoyances from my perspective are: 


Those engines that think they are "in charge" or are overly "helpful." (remember Microsoft's Clippy?). An example of what I mean by this is ones that define their own hot keys to switch back and forth between the primary keyboard and whatever "foreign" one the engine supports. In one case I found that, in order to return to English/Latin, I was unable to use my own ibus hotkey once under control of the engine and had to go to the menu to do so. Picky, perhaps, but annoying. 


Engines that don't permit "easy" (I know, a relative term) changes to things like display fonts, icons, and so forth. This may seem frivolous on the surface, but many (most?) database administrators have no familiarity whatever with other scripts and cannot, for instance, easily recognize the difference between Korean Hangul and Chinese Han script, much less tell at a glance what language a given string contains. Nor do these folks have the time needed to learn the subtleties. Therefore I set them up with a set of custom (consistent but easily distinguishable) icons targeted to their own circumstances. 


Another example is font size issues in the engine interfaces: just as someone who is unfamiliar with English might have trouble determining whether they are looking at a g or a q (here it's easy, but with some fonts it isn't clear at all), asking a DBA in the middle of troubleshooting some obscure problem, and who speaks only English or some other western language, to distinguish between the Han 未 and 末 (relative length of horizontal strokes) or between 千and 干 (different top strokes) at SMALL FONT SIZES is time-consuming at best. How about determining which Thai diacritic is in use: อื or อี? And trying to confirm/match the variety of Arabic letter forms against actual bit patterns on the disk ... 



Now I realize that I/T is not a common use case but, as the prevalence of multi-lingual data stores increases, it is becoming an issue. 


So my intent was not to disparage any particular engine, but to emphasize the importance of what I suppose are user-interface considerations, not only internally within each engine, but in conjunction with all engines in use. Getting the correct letter or symbol from a given key is obviously critical, but being able to clearly see and confirm each step in the process is also critical for system maintenance. 


So perhaps I should not have said "better than the replacements" but "more flexible than the replacements" or even "more easily integrated with other engines." 




----- Original Message -----

**Comment by mike-fabian** [comment]
If you use so many languages, ibus-typing-booster might be useful for you, combined with ibus-libpinyin for Chinese and ibus-anthy for Japanese (ibus-typing-booster does not support Chinese and Japanese, but for the other languages you can use ibus-typing-booster)

> I was unable to use my own ibus hotkey once under control of the engine and had to go to the menu to do so.

The keybindings in ibus-typing-booster are configurable, so if you changed the ibus hotkey form the default Super+Space to something else and it conflicts with any hotkey in ibus-typing-booster, you can change that hotkey in ibus-typing-booster to something else easily.

**Comment by mike-fabian** [comment]
> So perhaps I should not have said "better than the replacements" but "more flexible than the replacements" or even "more easily integrated with other engines."

I was mostly asking because I would consider not blacklisting some of the blacklisted engines anymore if there are good reasons why they might be useful.

**Comment by mike-fabian** [comment]
I am closing this as there was no further response.

**Comment by foberle** [comment]
Hi Mike - I sort of see your point, and should have responded earlier, but as long as I can "un-comment" the blacklist as needed and the files are not actually removed from the distros, that should still work. It's really difficult answering smart and otherwise experienced folks who say things like "why is this stuff so complicated, inconsistent, and so completely undocumented?" I'm reminded of Admiral Grace Hopper's comment about the benefits of having so many standards to choose from. I'm fairly convinced that if I hadn't been fiddling with multilingual data for almost 40 years, I'd never get any of it to work.

**Comment by mike-fabian** [comment]
You can repopen this if you like and tell me what not to blacklist and why.

Have you tried ibus-typing-booster? I guess it is better for you anyway than ibus-m17n.

See: https://github.com/ibus/ibus-m17n/issues/25#issuecomment-744069035

ibus-typing-booster doesn’t blacklist any of the m17n-db input methods.

**Comment by foberle** [comment]
Hi Mike - 


I got pulled away due to several issues back when I first posted my inquiry, so completely forgot that you had mentioned the ibus-typing-booster then. At any rate, I just downloaded it on to a Mint 20.2-Cinnamon machine to test, though I'll have to find some time to go through the documentation on how to use it. Given a few medical issues (getting old sucks by the way), that may be delayed for a bit, but I'll make sure to see how it fits in with the sorts of writing I do. 


Thanks for the response, and have a great day. 


Frank 

----- Original Message -----
